core/engine/src/value/mod.rs

```diff
@@ -242,45 +328,61 @@ impl JsValue {
     #[inline]
     #[must_use]
     pub const fn is_symbol(&self) -> bool {
-        matches!(self, Self::Symbol(_))
+        matches!(self.inner, InnerValue::Symbol(_))
     }
 
     /// Returns the symbol if the value is a symbol, otherwise `None`.
     #[inline]
     #[must_use]
     pub fn as_symbol(&self) -> Option<JsSymbol> {
-        match self {
-            Self::Symbol(symbol) => Some(symbol.clone()),
-            _ => None,
+        if let InnerValue::Symbol(symbol) = &self.inner {
+            Some(symbol.clone())
+        } else {
+            None
         }
     }
 
     /// Returns true if the value is undefined.
     #[inline]
     #[must_use]
     pub const fn is_undefined(&self) -> bool {
-        matches!(self, Self::Undefined)
+        matches!(&self.inner, InnerValue::Undefined)
     }
 
-    /// Returns true if the value is null.
+    /// Returns `()` if the value is undefined, otherwise `None`.
     #[inline]
     #[must_use]
-    pub const fn is_null(&self) -> bool {
-        matches!(self, Self::Null)
+    pub const fn as_undefined(&self) -> Option<()> {
+        if self.is_undefined() {
+            Some(())
+        } else {
+            None
+        }
     }
 
-    /// Returns true if the value is null or undefined.
+    /// Returns `Some(self)` if the value is defined, otherwise `None`.
     #[inline]
     #[must_use]
-    pub const fn is_null_or_undefined(&self) -> bool {
-        matches!(self, Self::Null | Self::Undefined)
+    pub const fn as_defined(&self) -> Option<&Self> {
+        if self.is_undefined() {
+            None
+        } else {
+            Some(self)
+        }
```

Comment:
This is rather useful (the same way `.ok_or()` is useful). I could remake this API into `defined_or()` and `defined_or_else()` which would cover the cases where it is used. What do you think?

Reply:
Would `defined_map` not be the same as the current `map` function that we already have? And `defined_or` would be `map_or`? We could add `map_or_else` to keep the same naming.

Looking at the places where you used `as_defined` I would say the ones in `typed_array` are probably better off with a `variant` match, since otherwise we match the value variant multiple times (not sure how well that is optimized away).

Reply:
Yeah, `map` was added recently was it not? I missed it when looking for it.

Reply:
I'm in the process of remove `as_defined()`. There's no value added vs `map`.